src/ipld/util.rs

```diff
@@ -437,3 +438,259 @@ impl<DB: Blockstore, T: Iterator<Item = Tipset> + Unpin> Stream for ChainStream<
         }
     }
 }
+
+enum UnorderedTask {
+    Cid(Cid),
+}
+
+pin_project! {
+    pub struct UnorderedChainStream<DB, T> {
+        tipset_iter: T,
+        db: Arc<DB>,
+        seen: Arc<Mutex<CidHashSet>>,
+        worker_handle: JoinHandle<anyhow::Result<()>>,
+        block_receiver: kanal::Receiver<anyhow::Result<Block>>,
+        extract_sender: kanal::Sender<Cid>,
+        stateroot_limit: ChainEpoch,
+        queue: Vec<UnorderedTask>,
+        fail_on_dead_links: bool,
+    }
+}
+
+impl<DB, T> UnorderedChainStream<DB, T> {
+    pub fn into_seen(self) -> CidHashSet {
+        match Arc::try_unwrap(self.seen) {
+            Ok(v) => v.into_inner(),
+            Err(v) => v.lock().clone(),
+        }
+    }
+}
+
+/// Stream all blocks that are reachable before the `stateroot_limit` epoch in an unordered fashion.
+/// After this limit, only block headers are streamed. Any dead links are reported as errors.
+///
+/// # Arguments
+///
+/// * `db` - A database that implements [`Blockstore`] interface.
+/// * `tipset_iter` - An iterator of [`Tipset`], descending order `$child -> $parent`.
+/// * `stateroot_limit` - An epoch that signifies how far back we need to inspect tipsets.
+/// in-depth. This has to be pre-calculated using this formula: `$cur_epoch - $depth`, where
```

Comment:
I don't see this in this block. Is this an artifact from a previous iteration of this PR?

Reply:
Not sure what the question is. The calculation happens on the client side, not anywhere in this implementation. Then we check epochs to avoid traversing blocks that we did not request. Does that make sense?

Reply:
Is there a misplaced period after `tipsets` and before the newline starting with `in-depth`?

Reply:
yes!